src/algebra/group/hom_instances.lean

```diff
@@ -0,0 +1,109 @@
+/-
+Copyright (c) 2018 Patrick Massot. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Patrick Massot, Kevin Buzzard, Scott Morrison, Johan Commelin, Chris Hughes,
+  Johannes Hölzl, Yury Kudryashov
+-/
+
+import algebra.group_power.basic
+
+/-!
+# instances on spaces of monoid and group morphisms
+
+We endow the space of monoid morphisms `M →* N` with a `comm_monoid` structure when the target is
+commutative, through pointwise multipliplication, and with a `comm_group` structure when the target
+is a commutative group. We also prove the same instances for additive situations.
+-/
+
+universes uM uN uP
+variables {M : Type uM} {N : Type uN} {P : Type uP}
+
+lemma nat.succ_eq_one_add (n : ℕ) : n.succ = 1 + n :=
+by rw [nat.succ_eq_add_one, nat.add_comm]
+
+/-- `(M →* N)` is a `comm_monoid` if `N` is commutative. -/
+instance [mul_one_class M] [comm_monoid N] : comm_monoid (M →* N) :=
+{ mul := (*),
+  mul_assoc := by intros; ext; apply mul_assoc,
+  one := 1,
+  one_mul := by intros; ext; apply one_mul,
+  mul_one := by intros; ext; apply mul_one,
+  mul_comm := by intros; ext; apply mul_comm,
+  npow := λ n f,
+  { to_fun := λ x, npow n (f x),
+    map_one' := by simp,
+    map_mul' := λ x y, by simp [mul_pow] },
+  npow_zero' := λ f, by { ext x, simp },
+  npow_succ' := λ n f, by { ext x, simp [pow_succ] } }
+
+/-- `(M →+ N)` is an `add_comm_monoid` if `N` is commutative. -/
+instance [add_zero_class M] [add_comm_monoid N] : add_comm_monoid (M →+ N) :=
+{ add := (+),
+  add_assoc := by intros; ext; apply add_assoc,
+  zero := 0,
+  zero_add := by intros; ext; apply zero_add,
+  add_zero := by intros; ext; apply add_zero,
+  add_comm := by intros; ext; apply add_comm,
+  nsmul := λ n f,
+  { to_fun := λ x, nsmul n (f x),
+    map_zero' := by simp [nsmul_zero],
+    map_add' := λ x y, by simp [nsmul_add] },
+  nsmul_zero' := λ f, by { ext x, simp [zero_nsmul], },
+  nsmul_succ' := λ n f, by { ext x, simp [nat.succ_eq_one_add, add_nsmul] } }
+
+attribute [to_additive] monoid_hom.comm_monoid
```

Comment:
I assume this has to be separate because the `nsmul_zero'` field confuses to_additive?

Reply:
This has to be separate because the proofs use lemmas phrased for `smul` and `pow`, and one can not set a `to_additive` between them because `pow` is taking its variables in the wrong order.